src/identity-utils.ts

```diff
@@ -0,0 +1,259 @@
+        for (let key in userIdentities) {
+            if (key === DEVICE_APPLICATION_STAMP) {
+                continue;
+            } else {
+                userIDArray[Types.IdentityType.getIdentityType(key)] =
+                    userIdentities[key];
+            }
+        }
+
+        concatenatedIdentities = userIDArray.reduce(
+            (prevValue: string, currentValue: string, index: number) => {
+                const idName: string = Types.IdentityType.getIdentityName(index);
+                return `${prevValue}${idName}=${currentValue};`;
+            },
+            cacheKey
+        );
+    }
+
+    return concatenatedIdentities;
+};
+
+export const hasValidCachedIdentity = (
+    method: IdentityAPIMethod,
+    proposedUserIdentities: IKnownIdentities,
+    idCache?: BaseVault<Dictionary<ICachedIdentityCall>>
+): boolean => {
+    // There is an edge case where multiple identity calls are taking place 
+    // before identify fires, so there may not be a cache.  See what happens when 
+    // the ? in idCache is removed to the following test
+    // "queued events contain login mpid instead of identify mpid when calling 
+    // login immediately after mParticle initializes"
+    const cache = idCache?.retrieve();
+
+    // if there is no cache, then there is no valid cached identity
+    if (!cache) {
+        return false;
+    }
+
+    const cacheKey: string = concatenateIdentities(
+        method,
+        proposedUserIdentities
+    );
+    const hashedKey = generateHash(cacheKey);
+
+    // if cache doesn't have the cacheKey, there is no valid cached identity
+    if (!cache.hasOwnProperty(hashedKey)) {
+        return false;
+    }
+    
+    // If there is a valid cache key, compare the expireTimestamp to the current time.
+    // If the current time is greater than the expireTimestamp, it is not a valid
+    // cached identity.
+    const expireTimestamp = cache[hashedKey].expireTimestamp;
+    
+    if (expireTimestamp < new Date().getTime()) {
+        return false;
+    } else {
+        return true;
+    }
+};
+
+export const getCachedIdentity = (
+    method: IdentityAPIMethod,
+    proposedUserIdentities: IKnownIdentities,
+    idCache: BaseVault<Dictionary<ICachedIdentityCall>>
+): ICachedIdentityCall | null => {
+    const cacheKey: string = concatenateIdentities(
+        method,
+        proposedUserIdentities
+    );
+    const hashedKey = generateHash(cacheKey);
+
+    const cache = idCache.retrieve();
+    const cachedIdentity = cache ? cache[hashedKey] : null;
+
+    return cachedIdentity;
+};
+
+// https://go.mparticle.com/work/SQDSDKS-6079
+export const createKnownIdentities = (
+    identityApiData: IdentityApiData,
+    deviceId: string
+): IKnownIdentities => {
+    const identitiesResult: IKnownIdentities = {};
+
+    if (
+        identityApiData?.userIdentities &&
+        isObject(identityApiData.userIdentities)
+    ) {
+        for (let identity in identityApiData.userIdentities) {
```

Comment:
I think this can be reduced to a single expression.

```suggestion
    if (isObject(identityApiData?.userIdentities)) {
        for (let identity in identityApiData.userIdentities) {
```